File: src/backend/agentchat/schema/knowledge.py
from typing import Optional

from pydantic import BaseModel, Field


class KnowledgeModelRefs(BaseModel):
    text_embedding_model_id: Optional[str] = Field(default=None, min_length=1)
    vl_embedding_model_id: Optional[str] = Field(default=None, min_length=1)
    rerank_model_id: Optional[str] = Field(default=None, min_length=1)


class KnowledgeModelRefsPatch(BaseModel):
    text_embedding_model_id: Optional[str] = Field(default=None, min_length=1)
    vl_embedding_model_id: Optional[str] = Field(default=None, min_length=1)
    rerank_model_id: Optional[str] = Field(default=None, min_length=1)


class KnowledgeIndexSettings(BaseModel):
    chunk_mode: str = Field(default="general", pattern="^(general|parent_child|qa)$")
    chunk_size: int = Field(default=1024, ge=1, le=8192)
    overlap: int = Field(default=120, ge=0, le=4096)
    separator: str = Field(default="\n\n", max_length=64)
    replace_consecutive_spaces: bool = Field(default=True)
    remove_urls_emails: bool = Field(default=False)
    image_indexing_mode: str = Field(default="dual", pattern="^(text_only|vl_only|dual)$")
    vector_backend: str = Field(default="milvus", pattern="^(milvus|chroma|milvus_lite)$")
    index_version: str = Field(default="v1", min_length=1, max_length=64)
    status: str = Field(default="active", pattern="^(active|disabled|archived)$")
    health_status: str = Field(default="ready", pattern="^(ready|degraded|stale|failed|unavailable)$")


class KnowledgeIndexSettingsPatch(BaseModel):
    chunk_mode: Optional[str] = Field(default=None, pattern="^(general|parent_child|qa)$")
    chunk_size: Optional[int] = Field(default=None, ge=1, le=8192)
    overlap: Optional[int] = Field(default=None, ge=0, le=4096)
    separator: Optional[str] = Field(default=None, max_length=64)
    replace_consecutive_spaces: Optional[bool] = Field(default=None)
    remove_urls_emails: Optional[bool] = Field(default=None)
    image_indexing_mode: Optional[str] = Field(default=None, pattern="^(text_only|vl_only|dual)$")
    vector_backend: Optional[str] = Field(default=None, pattern="^(milvus|chroma|milvus_lite)$")
    index_version: Optional[str] = Field(default=None, min_length=1, max_length=64)
    status: Optional[str] = Field(default=None, pattern="^(active|disabled|archived)$")
    health_status: Optional[str] = Field(default=None, pattern="^(ready|degraded|stale|failed|unavailable)$")


class KnowledgeGraphIndexSettings(BaseModel):
    entity_extraction_mode: str = Field(default="rule_llm", pattern="^(rule|llm|rule_llm)$")
    relation_schema: str = Field(default="open", pattern="^(open|typed)$")
    entity_normalization: bool = Field(default=True)
    evidence_backlink: bool = Field(default=True)
    use_rag_entry_chunk: bool = Field(default=True)
    index_version: str = Field(default="v1", min_length=1, max_length=64)
    health_status: str = Field(default="ready", pattern="^(ready|degraded|stale|failed|unavailable)$")


class KnowledgeGraphIndexSettingsPatch(BaseModel):
    entity_extraction_mode: Optional[str] = Field(default=None, pattern="^(rule|llm|rule_llm)$")
    relation_schema: Optional[str] = Field(default=None, pattern="^(open|typed)$")
    entity_normalization: Optional[bool] = Field(default=None)
    evidence_backlink: Optional[bool] = Field(default=None)
    use_rag_entry_chunk: Optional[bool] = Field(default=None)
    index_version: Optional[str] = Field(default=None, min_length=1, max_length=64)
    health_status: Optional[str] = Field(default=None, pattern="^(ready|degraded|stale|failed|unavailable)$")


class KnowledgeRetrievalSettings(BaseModel):
    default_mode: str = Field(default="rag", pattern="^(auto|hybrid|rag|graphrag|rag_graph)$")
    refill_policy: str = Field(default="smart", pattern="^(none|auto|smart)$")
    top_k: int = Field(default=5, ge=1, le=50)
    rerank_enabled: bool = Field(default=True)
    rerank_top_k: int = Field(default=4, ge=1, le=50)
    score_threshold: Optional[float] = Field(default=None, ge=0, le=1)
    graph_hop_limit: int = Field(default=2, ge=1, le=3)
    max_paths_per_entity: int = Field(default=5, ge=1, le=20)


class KnowledgeRetrievalSettingsPatch(BaseModel):
    default_mode: Optional[str] = Field(default=None, pattern="^(auto|hybrid|rag|graphrag|rag_graph)$")
    refill_policy: Optional[str] = Field(default=None, pattern="^(none|auto|smart)$")
    top_k: Optional[int] = Field(default=None, ge=1, le=50)
    rerank_enabled: Optional[bool] = Field(default=None)
    rerank_top_k: Optional[int] = Field(default=None, ge=1, le=50)
    score_threshold: Optional[float] = Field(default=None, ge=0, le=1)
    graph_hop_limit: Optional[int] = Field(default=None, ge=1, le=3)
    max_paths_per_entity: Optional[int] = Field(default=None, ge=1, le=20)


class KnowledgeConfig(BaseModel):
    index_capability: str = Field(default="rag", pattern="^(rag|rag_graph)$")
    domain_pack_id: Optional[str] = Field(default=None, min_length=1, max_length=64)
    model_refs: KnowledgeModelRefs = Field(default_factory=KnowledgeModelRefs)
    index_settings: KnowledgeIndexSettings = Field(default_factory=KnowledgeIndexSettings)
    graph_index_settings: KnowledgeGraphIndexSettings = Field(default_factory=KnowledgeGraphIndexSettings)
    retrieval_settings: KnowledgeRetrievalSettings = Field(default_factory=KnowledgeRetrievalSettings)


class KnowledgeConfigPatch(BaseModel):
    index_capability: Optional[str] = Field(default=None, pattern="^(rag|rag_graph)$")
    domain_pack_id: Optional[str] = Field(default=None, min_length=1, max_length=64)
    model_refs: Optional[KnowledgeModelRefsPatch] = Field(default=None)
    index_settings: Optional[KnowledgeIndexSettingsPatch] = Field(default=None)
    graph_index_settings: Optional[KnowledgeGraphIndexSettingsPatch] = Field(default=None)
    retrieval_settings: Optional[KnowledgeRetrievalSettingsPatch] = Field(default=None)


class KnowledgeCreateRequest(BaseModel):
    knowledge_name: str = Field(description="knowledge name", min_length=2, max_length=10)
    knowledge_desc: Optional[str] = Field(
        description="knowledge description",
        min_length=10,
        max_length=200,
        default=None,
    )
    knowledge_config: Optional[KnowledgeConfig] = Field(default=None)


class KnowledgeUpdateRequest(BaseModel):
    knowledge_id: str = Field(description="knowledge id")
    knowledge_name: Optional[str] = Field(default=None, description="knowledge name", min_length=2, max_length=10)
    knowledge_desc: Optional[str] = Field(
        default=None,
        description="knowledge description",
        min_length=10,
        max_length=200,
    )
    knowledge_config: Optional[KnowledgeConfigPatch] = Field(default=None)

File: apps/web/src/apis/knowledge.ts
import { request } from '../utils/request'

export interface UnifiedResponse<T = any> {
  status_code: number
  status_message: string
  data?: T
}

export interface KnowledgeConfigPayload {
  index_capability: 'rag' | 'rag_graph'
  graphrag_project_id: string | null
  domain_pack_id: string | null
  eval_profile_id: string | null
  model_refs: {
    text_embedding_model_id: string | null
    vl_embedding_model_id: string | null
    rerank_model_id: string | null
  }
  index_settings: {
    chunk_mode: 'general' | 'parent_child' | 'qa'
    chunk_size: number
    overlap: number
    separator: string
    replace_consecutive_spaces: boolean
    remove_urls_emails: boolean
    image_indexing_mode: 'text_only' | 'vl_only' | 'dual'
    vector_backend: 'milvus' | 'chroma' | 'milvus_lite'
    index_version: string
    status: 'active' | 'disabled' | 'archived'
    health_status: 'ready' | 'degraded' | 'stale' | 'failed' | 'unavailable'
  }
  graph_index_settings: {
    entity_extraction_mode: 'rule' | 'llm' | 'rule_llm'
    relation_schema: 'open' | 'typed'
    entity_normalization: boolean
    evidence_backlink: boolean
    use_rag_entry_chunk: boolean
    community_report_prompt_id: string | null
    index_version: string
    health_status: 'ready' | 'degraded' | 'stale' | 'failed' | 'unavailable'
    graph_index_status?: 'ready' | 'queued' | 'running' | 'stale' | 'failed' | 'unavailable' | 'not_built'
    community_detection_status?: 'ready' | 'queued' | 'running' | 'stale' | 'failed' | 'unavailable' | 'not_built'
    community_report_status?: 'ready' | 'queued' | 'running' | 'stale' | 'failed' | 'unavailable' | 'not_built'
    community_version?: string
  }
  retrieval_settings: {
    default_mode: 'auto' | 'hybrid' | 'rag' | 'graphrag' | 'rag_graph' | 'rag_graph_deep'
    profile: string
    refill_policy: 'none' | 'auto' | 'smart'
    top_k: number
    rerank_enabled: boolean
    rerank_top_k: number
    score_threshold: number | null
    graph_hop_limit: number
    max_paths_per_entity: number
  }
}

export interface KnowledgeConfigPatchPayload {
  index_capability?: KnowledgeConfigPayload['index_capability']
  graphrag_project_id?: string | null
  domain_pack_id?: string | null
  eval_profile_id?: string | null
  model_refs?: Partial<KnowledgeConfigPayload['model_refs']>
  index_settings?: Partial<KnowledgeConfigPayload['index_settings']>
  graph_index_settings?: Partial<KnowledgeConfigPayload['graph_index_settings']>
  retrieval_settings?: Partial<KnowledgeConfigPayload['retrieval_settings']>
}

export interface KnowledgeResponse {
  id: string
  name: string
  description: string | null
  user_id: string | null
  create_time: string
  update_time: string
  count: number
  file_size: string
  processing_count?: number
  failed_count?: number
  completed_count?: number
  knowledge_config?: KnowledgeConfigPayload
}

export interface KnowledgeCreateRequest {
  knowledge_name: string
  knowledge_desc?: string
  knowledge_config?: KnowledgeConfigPayload
}

export interface KnowledgeUpdateRequest {
  knowledge_id: string
  knowledge_name?: string
  knowledge_desc?: string
  knowledge_config?: KnowledgeConfigPatchPayload
}

export interface KnowledgeDeleteRequest {
  knowledge_id: string
}

export type KnowledgeReindexAction =
  | 'text_index'
  | 'image_index'
  | 'bm25_index'
  | 'graph_index'
  | 'community_detection'
  | 'community_report'
  | 'full_rebuild'

export interface KnowledgeConfigImpactResponse {
  changed_fields: string[]
  immediate_effect_fields: string[]
  text_reindex_required: boolean
  image_reindex_required: boolean
  bm25_reindex_required: boolean
  graph_update_required: boolean
  community_detection_required: boolean
  community_report_required: boolean
  full_rebuild_required: boolean
  recommended_action: KnowledgeReindexAction | 'save_only'
}

export interface KnowledgeModelBindingPayload {
  llm_id: string
  model: string
  provider: string
}

interface KnowledgeRetrievalRequest {
  query: string
  knowledge_id: string | string[]
  top_k?: number
  retrieval_mode?: string
}

interface KnowledgeRetrievalRound {
  round: number
  mode: string
  query: string
  trigger: string
  quality_reason: string | null
  document_count?: number | null
  top_score?: number | null
  score_threshold?: number | null
  path_count?: number | null
  entity_count?: number | null
  content_found: boolean
}

interface KnowledgeRetrievalResponse {
  content: string
  actual_mode: string
  first_mode: string
  final_mode: string
  second_pass_used: boolean
  fallback_triggered: boolean
  fallback_reason: string | null
  round_count: number
  metadata?: {
    first_mode: string
    final_mode: string
    second_pass_used: boolean
    fallback_triggered: boolean
    fallback_reason: string | null
    final_quality_reason?: string | null
    round_count: number
    rounds: KnowledgeRetrievalRound[]
    query_variants: string[]
    rewritten_query_used: boolean
  }
}

export function getKnowledgeListAPI() {
  return request<UnifiedResponse<KnowledgeResponse[]>>({
    url: '/api/v1/knowledge/select',
    method: 'GET',
  })
}

export function createKnowledgeAPI(data: KnowledgeCreateRequest) {
  return request<UnifiedResponse<KnowledgeResponse>>({
    url: '/api/v1/knowledge/create',
    method: 'POST',
    data,
  })
}

export function updateKnowledgeAPI(data: KnowledgeUpdateRequest) {
  return request<UnifiedResponse<null>>({
    url: '/api/v1/knowledge/update',
    method: 'PUT',
    data,
  })
}

export function getKnowledgeConfigAPI(knowledgeId: string) {
  return request<UnifiedResponse<KnowledgeConfigPayload>>({
    url: `/api/v1/knowledge/${knowledgeId}/config`,
    method: 'GET',
  })
}

export function updateKnowledgeConfigAPI(knowledgeId: string, nextConfig: KnowledgeConfigPayload) {
  return request<UnifiedResponse<null>>({
    url: `/api/v1/knowledge/${knowledgeId}/config`,
    method: 'PUT',
    data: {
      next_config: nextConfig,
    },
  })
}

export function analyzeKnowledgeConfigImpactAPI(knowledgeId: string, nextConfig: KnowledgeConfigPayload) {
  return request<UnifiedResponse<KnowledgeConfigImpactResponse>>({
    url: `/api/v1/knowledge/${knowledgeId}/config/impact`,
    method: 'POST',
    data: {
      next_config: nextConfig,
    },
  })
}

export function runKnowledgeReindexActionAPI(knowledgeId: string, action: KnowledgeReindexAction) {
  return request<UnifiedResponse<{ knowledge_id: string; action: KnowledgeReindexAction; status: string }>>({
    url: `/api/v1/knowledge/${knowledgeId}/reindex/${action}`,
    method: 'POST',
  })
}

export function deleteKnowledgeAPI(data: KnowledgeDeleteRequest) {
  return request<UnifiedResponse<null>>({
    url: '/api/v1/knowledge/delete',
    method: 'DELETE',
    data,
  })
}
